openpgp/ecdh/ecdh.go

```diff
@@ -0,0 +1,240 @@
+package ecdh
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/aes"
+	"crypto/elliptic"
+	"encoding/binary"
+	"errors"
+	"io"
+	"math/big"
+)
+
+type PublicKey struct {
+	elliptic.Curve
+	X, Y *big.Int
+}
+
+type PrivateKey struct {
+	PublicKey
+	X *big.Int
+}
+
+func RandomScalar(random io.Reader, N *big.Int) (*big.Int, error) {
+	randLen := new(big.Int).Sub(N, big.NewInt(2)).BitLen()
+	randBuf := make([]byte, randLen/8)
+	err := nonZeroRandomBytes(randBuf, random)
+	if err != nil {
+		return nil, err
+	}
+
+	m := new(big.Int).SetBytes(randBuf)
+	m = new(big.Int).Add(m, big.NewInt(1))
+
+	return m, nil
+}
+
+func (e *PublicKey) Kdf(S []byte, kdf_params []byte, hash crypto.Hash) []byte {
+	S_len := (e.Curve.Params().P.BitLen() + 7) / 8
+	buf := new(bytes.Buffer)
+	buf.Write([]byte{0, 0, 0, 1})
+	if S_len > len(S) {
+		// If we got invalid S (bigger than curve's P), we are going
+		// to produce invalid key. Garbage in, garbage out.
+		buf.Write(make([]byte, S_len-len(S)))
+	}
+	buf.Write(S)
+	buf.Write(kdf_params)
+
+	hashw := hash.New()
+
+	hashw.Write(buf.Bytes())
+	key := hashw.Sum(nil)
+
+	return key
+}
+
```

Comment:
Oh, there is a comment about RFC but in the next function, I think I must have reordered them. But I will leave more comments about this.

It's a little bit different than what KBPGP does, as it does not split the buffer into 64bit words first, it operates on the byte buffer. But it's essentially the non-shifting algorithm from the RFC.